File: ALM-Model-Lab/src/math/rates/hwPricing.ts
/**
 * Hull-White 1-factor analytical pricing — TypeScript port of `research/hw_pricing.py`.
 *
 * Closed forms for the shifted-Gaussian HW1F model:
 *
 *   r(t) = f^M(0, t) + X(t),     dX = -a X dt + sigma dW.
 *
 * Bachelier-equivalent normal vol of a caplet (Andersen-Piterbarg Vol II Sec 10.1.6):
 *
 *   sigma_N(T, tau_u)^2 * T = sigma^2 * B(tau_u)^2 * (1 - exp(-2 a T))/(2 a)
 *                              * ((1 + F * tau_u) / tau_u)^2
 *
 * with B(tau) = (1 - exp(-a tau)) / a.
 */

import type { ZeroCurve } from "./bootstrap";
import { buildSwap } from "./bgmPricing";

const SMALL_A_THRESHOLD = 1e-10;

/** B(tau) = (1 - exp(-a*tau)) / a, with small-a expansion. */
export function bFunction(a: number, tau: number): number {
  if (Math.abs(a) < SMALL_A_THRESHOLD) {
    return tau - 0.5 * a * tau * tau;
  }
  return (1.0 - Math.exp(-a * tau)) / a;
}

/**
 * Reconstructed P(t, t+tau) under HW given the latent factor X(t).
 *
 * Brigo-Mercurio Eq 3.39-3.40 in shifted form:
 *   P(t, T) = (P^M(0, T) / P^M(0, t)) * exp(-B(tau) * X(t) - V(t, tau))
 * with V(t, tau) = (sigma^2 / (4 a)) * (1 - exp(-2 a t)) * B(tau)^2.
 */
export function hwBondPrice(
  P0_t: number,
  P0_T: number,
  a: number,
  sigma: number,
  t: number,
  tau: number,
  xT: number,
): number {
  const B = bFunction(a, tau);
  const V = (sigma * sigma / (4.0 * a)) * (1.0 - Math.exp(-2.0 * a * t)) * B * B;
  return (P0_T / P0_t) * Math.exp(-B * xT - V);
}

/**
 * Model normal (Bachelier) vol of a caplet under HW.
 *
 * @param a HW mean-reversion.
 * @param sigma HW volatility.
 * @param expiry Caplet expiry T in years.
 * @param underlyingTau Length of rate accrual period tau_u in years.
 * @param forwardRate Initial simply-compounded forward F(0, T, T+tau_u).
 */
export function hwCapletNormalVol(
  a: number,
  sigma: number,
  expiry: number,
  underlyingTau: number,
  forwardRate: number,
): number {
  if (expiry <= 0) return 0;
  const Bu = bFunction(a, underlyingTau);
  const integrated = (sigma * sigma * (1.0 - Math.exp(-2.0 * a * expiry))) / (2.0 * a);
  const coef = ((1.0 + forwardRate * underlyingTau) / underlyingTau) ** 2;
  const sigmaNSqT = integrated * (Bu * Bu) * coef;
  return Math.sqrt(Math.max(sigmaNSqT, 0) / expiry);
}

/**
 * ATM swaption normal (Bachelier) vol under HW1F — Gaussian-factor freeze.
 *
 * At expiry T_alpha bond prices are P(T_alpha, T_k) = A_k exp(-B(T_k - T_alpha) X),
 * so the par swap rate is a smooth function of the single factor. Linearising
 * at X = 0 with forward bond prices P_k ≈ P(0, T_k) / P(0, T_alpha):
 *
 *   dS/dX = [ B(T_beta - T_alpha) P_beta A + (1 - P_beta) Σ_k τ_k B(T_k - T_alpha) P_k ] / A²
 *   σ_N² T_alpha = (dS/dX)² σ² (1 - e^{-2 a T_alpha}) / (2a)
 *
 * The one-period limit reproduces hwCapletNormalVol exactly (same freeze:
 * dS/dX collapses to B(τ)(1 + S τ_s)/τ_s). Swap schedule conventions (annual
 * ACT/360) come from buildSwap so the result is directly comparable to the
 * quoted ATM normal vols and the BGM fit.
 */
export function hwSwaptionNormalVol(
  curve: ZeroCurve,
  a: number,
  sigma: number,
  TAlpha: number,
  TBeta: number,
): number {
  if (TAlpha <= 0) return 0;
  const swap = buildSwap(curve, TAlpha, TBeta);
  const dfA = curve.discountFactor(TAlpha);
  // Bond prices as seen from T_alpha (X = 0 freeze).
  const pFwd = swap.paymentDates.map((t) => curve.discountFactor(t) / dfA);
  const annuity = swap.forwardTaus.reduce((s, tau, i) => s + tau * pFwd[i], 0);
  const pBeta = pFwd[pFwd.length - 1];
  const dNum = bFunction(a, TBeta - TAlpha) * pBeta;
  const dAnn = swap.forwardTaus.reduce(
    (s, tau, i) => s + tau * bFunction(a, swap.paymentDates[i] - TAlpha) * pFwd[i],
    0,
  );
  const dSdX = (dNum * annuity + (1.0 - pBeta) * dAnn) / (annuity * annuity);
  const integrated =
    Math.abs(a) < SMALL_A_THRESHOLD
      ? sigma * sigma * TAlpha
      : (sigma * sigma * (1.0 - Math.exp(-2.0 * a * TAlpha))) / (2.0 * a);
  return Math.abs(dSdX) * Math.sqrt(integrated / TAlpha);
}

/** Vectorised over arrays of expiries and forwards (parallel arrays). */
export function hwCapletNormalVolVec(
  a: number,
  sigma: number,
  expiries: ArrayLike<number>,
  underlyingTau: number,
  forwards: ArrayLike<number>,
): number[] {
  const out = new Array<number>(expiries.length);
  for (let i = 0; i < expiries.length; i++) {
    out[i] = hwCapletNormalVol(a, sigma, expiries[i], underlyingTau, forwards[i]);
  }
  return out;
}

File: ALM-Model-Lab/research/hw_pricing.py
"""Hull-White 1-factor analytical pricing.

Closed forms for the shifted-Gaussian HW1F model:

    r(t) = f^M(0, t) + X(t),     dX = -a X dt + sigma dW

The simply-compounded forward rate F(t, T, T+tau_u) = (P(t,T)/P(t,T+tau_u) - 1)/tau_u
has, under the T+tau_u-forward measure, dynamics that yield a closed-form
Bachelier-equivalent normal vol used for caplet calibration.

Derivation (Andersen-Piterbarg Vol II Sec 10.1.6, freezing the F coefficient
in the lognormal-to-normal vol conversion):

    sigma_N^2(T, tau_u) * T = sigma^2 * B(tau_u)^2 * (1 - exp(-2 a T))/(2 a)
                                 * ((1 + F * tau_u) / tau_u)^2

with B(tau) = (1 - exp(-a tau)) / a.

Closed forms here also include analytical bond prices used by the simulator
to reconstruct yields from the latent X factor.
"""

from __future__ import annotations

import math
import numpy as np


# ---------------------------------------------------------------------------
# Building blocks
# ---------------------------------------------------------------------------

def b_function(a: float, tau: float | np.ndarray) -> float | np.ndarray:
    """B(tau) = (1 - exp(-a*tau)) / a.

    Time-homogeneous in HW: depends only on (T - t), not on t separately.
    The tau -> 0 limit is handled by series expansion to avoid 0/0.
    """
    a_arr = np.asarray(a, dtype=float)
    tau_arr = np.asarray(tau, dtype=float)
    # Small-a expansion: B ~ tau - 0.5 a tau^2 + ...
    if np.isscalar(a) and abs(a) < 1e-10:
        return tau_arr - 0.5 * a_arr * tau_arr * tau_arr
    return (1.0 - np.exp(-a_arr * tau_arr)) / a_arr


def hw_bond_price(
    P0_t: float,
    P0_T: float,
    a: float,
    sigma: float,
    t: float,
    tau: float,
    x_t: float,
) -> float:
    """Reconstructed P(t, t+tau) under HW given the latent factor X(t).

    Brigo-Mercurio Eq 3.39-3.40 in shifted form:

        P(t, T) = (P^M(0, T) / P^M(0, t)) * exp(-B(tau) * X(t) - V(t, tau))

    with V(t, tau) = (sigma^2 / (4 a)) * (1 - exp(-2 a t)) * B(tau)^2.
    """
    B = float(b_function(a, tau))
    V = (sigma * sigma / (4.0 * a)) * (1.0 - math.exp(-2.0 * a * t)) * B * B
    return (P0_T / P0_t) * math.exp(-B * x_t - V)


# ---------------------------------------------------------------------------
# Caplet normal vol under HW (calibration target)
# ---------------------------------------------------------------------------

def hw_caplet_normal_vol(
    a: float,
    sigma: float,
    expiry: float,
    underlying_tau: float,
    forward_rate: float,
) -> float:
    """Model normal (Bachelier) vol of a caplet under HW.

    Parameters
    ----------
    a, sigma : float
        HW mean-reversion and volatility parameters.
    expiry : float
        Caplet expiry T in years.
    underlying_tau : float
        Length of the rate's accrual period tau_u in years (typical 0.25 for 3M SOFR).
    forward_rate : float
        Initial simply-compounded forward F(0, T, T+tau_u) (decimal).

    Returns
    -------
    float
        Implied Bachelier normal vol of F (decimal).
    """
    if expiry <= 0.0:
        return 0.0
    B_u = float(b_function(a, underlying_tau))
    # Variance contribution from sigma * dW integrated over [0, T] times the
    # rate-to-bond-vol conversion factor ((1+F*tau_u)/tau_u)^2.
    integrated = sigma * sigma * (1.0 - math.exp(-2.0 * a * expiry)) / (2.0 * a)
    coef = ((1.0 + forward_rate * underlying_tau) / underlying_tau) ** 2
    sigma_n_sq_T = integrated * (B_u * B_u) * coef
    return math.sqrt(max(sigma_n_sq_T, 0.0) / expiry)


def hw_caplet_normal_vol_vec(
    a: float,
    sigma: float,
    expiries: np.ndarray,
    underlying_tau: float,
    forwards: np.ndarray,
) -> np.ndarray:
    """Vectorised caplet normal vol over arrays of expiries and forwards."""
    expiries = np.asarray(expiries, dtype=float)
    forwards = np.asarray(forwards, dtype=float)
    B_u = float(b_function(a, underlying_tau))
    integrated = sigma * sigma * (1.0 - np.exp(-2.0 * a * expiries)) / (2.0 * a)
    coef = ((1.0 + forwards * underlying_tau) / underlying_tau) ** 2
    sigma_n_sq_T = integrated * (B_u * B_u) * coef
    return np.sqrt(np.maximum(sigma_n_sq_T, 0.0) / np.where(expiries > 0, expiries, 1.0))


# ---------------------------------------------------------------------------
# Bachelier caplet price (used for sanity / pricing in run reports)
# ---------------------------------------------------------------------------

def bachelier_caplet_price(
    forward: float,
    strike: float,
    expiry: float,
    underlying_tau: float,
    df_pay: float,
    sigma_n: float,
) -> float:
    """Bachelier (normal) caplet price.

    Pay-off: tau_u * max(F - K, 0) at T + tau_u; discounted by df_pay = DF(0, T+tau_u).
    """
    from math import sqrt, exp, pi
    from statistics import NormalDist

    if expiry <= 0.0 or sigma_n <= 0.0:
        return underlying_tau * df_pay * max(forward - strike, 0.0)
    sd = sigma_n * sqrt(expiry)
    d = (forward - strike) / sd
    nd = NormalDist().cdf(d)
    pdf_d = exp(-0.5 * d * d) / sqrt(2.0 * pi)
    return underlying_tau * df_pay * ((forward - strike) * nd + sd * pdf_d)


if __name__ == "__main__":
    # Smoke test: known-good a=0.05, sigma=0.01 should produce HW caplet normal vols
    # of order 80-120 bps for SOFR-like 3.6% forwards across 1-10Y expiries.
    a, sigma = 0.05, 0.01
    tau_u = 0.25
    F = 0.036
    print(f"HW pricing smoke test: a={a}, sigma={sigma}, F={F}, tau_u={tau_u}")
    for T in (1.0, 2.0, 3.0, 5.0, 10.0, 15.0):
        v = hw_caplet_normal_vol(a, sigma, T, tau_u, F)
        print(f"  T={T:5.1f}y  sigma_N={v*1e4:7.2f} bp")
